File: app/src/main/java/org/gnucash/android/db/MigrationHelper.java
/*
 * Copyright (c) 2014 - 2015 Ngewi Fet <ngewif@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gnucash.android.db;

import static android.database.DatabaseUtils.sqlEscapeString;
import static org.gnucash.android.db.DatabaseHelper.createResetBalancesTriggers;
import static org.gnucash.android.db.DatabaseSchema.AccountEntry;
import static org.gnucash.android.db.DatabaseSchema.BudgetAmountEntry;
import static org.gnucash.android.db.DatabaseSchema.CommodityEntry;
import static org.gnucash.android.db.DatabaseSchema.TransactionEntry;

import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;

import androidx.annotation.NonNull;

import org.gnucash.android.R;
import org.gnucash.android.importer.CommoditiesXmlHandler;
import org.gnucash.android.model.Commodity;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.XMLReader;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import timber.log.Timber;

/**
 * Collection of helper methods which are used during database migrations
 *
 * @author Ngewi Fet <ngewif@gmail.com>
 */
@SuppressWarnings("unused")
public class MigrationHelper {

    /**
     * Imports commodities into the database from XML resource file
     */
    static void importCommodities(@NonNull DatabaseHolder holder) throws SAXException, ParserConfigurationException, IOException {
        SAXParserFactory spf = SAXParserFactory.newInstance();
        SAXParser sp = spf.newSAXParser();
        XMLReader xr = sp.getXMLReader();

        InputStream commoditiesInputStream = holder.context.getResources()
            .openRawResource(R.raw.iso_4217_currencies);
        BufferedInputStream bos = new BufferedInputStream(commoditiesInputStream);

        /* Create handler to handle XML Tags ( extends DefaultHandler ) */
        CommoditiesXmlHandler handler = new CommoditiesXmlHandler(holder);

        xr.setContentHandler(handler);
        xr.parse(new InputSource(bos));
    }

    public static void migrate(@NonNull Context context, @NonNull SQLiteDatabase db, int oldVersion, int newVersion) {
        if (oldVersion < 16) {
            migrateTo16(db);
        }
        if (oldVersion < 17) {
            migrateTo17(db);
        }
        if (oldVersion < 18) {
            migrateTo18(db);
        }
        if (oldVersion < 19) {
            migrateTo19(db);
        }
        if ((oldVersion >= 19) && (oldVersion < 21)) {
            migrateTo21(db);
        }
        if (oldVersion < 23) {
            migrateTo23(context, db);
        }
    }

    /**
     * Upgrade the database to version 16.
     *
     * @param db the database.
     */
    private static void migrateTo16(SQLiteDatabase db) {
        Timber.i("Upgrading database to version 16");

        String sqlAddQuoteSource = "ALTER TABLE " + CommodityEntry.TABLE_NAME +
            " ADD COLUMN " + CommodityEntry.COLUMN_QUOTE_SOURCE + " varchar(255)";
        String sqlAddQuoteTZ = "ALTER TABLE " + CommodityEntry.TABLE_NAME +
            " ADD COLUMN " + CommodityEntry.COLUMN_QUOTE_TZ + " varchar(100)";

        db.execSQL(sqlAddQuoteSource);
        db.execSQL(sqlAddQuoteTZ);
    }

    /**
     * Upgrade the database to version 17.
     *
     * @param db the database.
     */
    private static void migrateTo17(SQLiteDatabase db) {
        Timber.i("Upgrading database to version 17");

        String sqlAddBudgetNotes = "ALTER TABLE " + BudgetAmountEntry.TABLE_NAME +
            " ADD COLUMN " + BudgetAmountEntry.COLUMN_NOTES + " text";

        db.execSQL(sqlAddBudgetNotes);
    }

    /**
     * Upgrade the database to version 18.
     *
     * @param db the database.
     */
    private static void migrateTo18(SQLiteDatabase db) {
        Timber.i("Upgrading database to version 18");

        String sqlAddNotes = "ALTER TABLE " + AccountEntry.TABLE_NAME
            + " ADD COLUMN " + AccountEntry.COLUMN_NOTES + " text";
        String sqlAddBalance = "ALTER TABLE " + AccountEntry.TABLE_NAME
            + " ADD COLUMN " + AccountEntry.COLUMN_BALANCE + " varchar(255)";
        String sqlAddClearedBalance = "ALTER TABLE " + AccountEntry.TABLE_NAME
            + " ADD COLUMN " + AccountEntry.COLUMN_CLEARED_BALANCE + " varchar(255)";
        String sqlAddNoClosingBalance = "ALTER TABLE " + AccountEntry.TABLE_NAME
            + " ADD COLUMN " + AccountEntry.COLUMN_NOCLOSING_BALANCE + " varchar(255)";
        String sqlAddReconciledBalance = "ALTER TABLE " + AccountEntry.TABLE_NAME
            + " ADD COLUMN " + AccountEntry.COLUMN_RECONCILED_BALANCE + " varchar(255)";

        db.execSQL(sqlAddNotes);
        db.execSQL(sqlAddBalance);
        db.execSQL(sqlAddClearedBalance);
        db.execSQL(sqlAddNoClosingBalance);
        db.execSQL(sqlAddReconciledBalance);
        createResetBalancesTriggers(db);
    }

    /**
     * Upgrade the database to version 19.
     *
     * @param db the database.
     */
    private static void migrateTo19(SQLiteDatabase db) {
        Timber.i("Upgrading database to version 19");

        // Fetch list of accounts with mismatched currencies.
        String sqlAccountCurrencyWrong = "SELECT DISTINCT a." + AccountEntry.COLUMN_CURRENCY + ", a." + AccountEntry.COLUMN_COMMODITY_UID + ", c." + CommodityEntry.COLUMN_UID
            + " FROM " + AccountEntry.TABLE_NAME + " a, " + CommodityEntry.TABLE_NAME + " c"
            + " WHERE a." + AccountEntry.COLUMN_CURRENCY + " = c." + CommodityEntry.COLUMN_MNEMONIC
            + " AND (c." + CommodityEntry.COLUMN_NAMESPACE + " = " + sqlEscapeString(Commodity.COMMODITY_CURRENCY)
            + " OR c." + CommodityEntry.COLUMN_NAMESPACE + " = " + sqlEscapeString(Commodity.COMMODITY_ISO4217) + ")"
            + " AND a." + AccountEntry.COLUMN_COMMODITY_UID + " != c." + CommodityEntry.COLUMN_UID;
        Cursor cursor = db.rawQuery(sqlAccountCurrencyWrong, null);
        List<AccountCurrency> accountsWrong = new ArrayList<>();
        if (cursor.moveToFirst()) {
            do {
                String currencyCode = cursor.getString(0);
                String commodityUIDOld = cursor.getString(1);
                String commodityUIDNew = cursor.getString(2);
                accountsWrong.add(new AccountCurrency(currencyCode, commodityUIDOld, commodityUIDNew));
            } while (cursor.moveToNext());
        }
        cursor.close();

        // Update with correct commodities.
        for (AccountCurrency accountWrong : accountsWrong) {
            String sql = "UPDATE " + AccountEntry.TABLE_NAME
                + " SET " + AccountEntry.COLUMN_COMMODITY_UID + " = " + sqlEscapeString(accountWrong.commodityUIDNew)
                + " WHERE " + AccountEntry.COLUMN_CURRENCY + " = " + sqlEscapeString(accountWrong.currencyCode)
                + " AND " + AccountEntry.COLUMN_COMMODITY_UID + " = " + sqlEscapeString(accountWrong.commodityUIDOld);
            db.execSQL(sql);
        }
    }

    private static class AccountCurrency {
        @NonNull
        public final String currencyCode;
        @NonNull
        public final String commodityUIDOld;
        @NonNull
        public final String commodityUIDNew;

        private AccountCurrency(
            @NonNull String currencyCode,
            @NonNull String commodityUIDOld,
            @NonNull String commodityUIDNew) {
            this.currencyCode = currencyCode;
            this.commodityUIDOld = commodityUIDOld;
            this.commodityUIDNew = commodityUIDNew;
        }
    }

    /**
     * Upgrade the database to version 21.
     *
     * @param db the database.
     */
    private static void migrateTo21(SQLiteDatabase db) {
        Timber.i("Upgrading database to version 21");

        Cursor cursor = db.rawQuery("PRAGMA table_info(" + AccountEntry.TABLE_NAME + ")", null);
        try {
            if (cursor.moveToFirst()) {
                final int indexName = cursor.getColumnIndex("name");
                do {
                    String name = cursor.getString(indexName);
                    if (AccountEntry.COLUMN_CURRENCY.equals(name)) {
                        return;
                    }
                } while (cursor.moveToNext());
            }
        } finally {
            cursor.close();
        }

        // Restore the currency code column that was deleted in v19.
        String sqlAccountCurrency = "ALTER TABLE " + AccountEntry.TABLE_NAME
            + " ADD COLUMN " + AccountEntry.COLUMN_CURRENCY + " varchar(255)";
        try {
            db.execSQL(sqlAccountCurrency);
        } catch (SQLException e) {
            Timber.e(e);
        }

        // Restore the currency code column.
        String sqlTransactionCurrency = "ALTER TABLE " + TransactionEntry.TABLE_NAME
            + " ADD COLUMN " + TransactionEntry.COLUMN_CURRENCY + " varchar(255)";
        try {
            db.execSQL(sqlTransactionCurrency);
        } catch (SQLException e) {
            Timber.e(e);
        }
    }

    /**
     * Upgrade the database to version 23.
     *
     * @param context the context.
     * @param db      the database.
     */
    private static void migrateTo23(@NonNull Context context, @NonNull SQLiteDatabase db) {
        Timber.i("Upgrading database to version 23");

        boolean hasColumnQuoteFlag = false;
        Cursor cursor = db.rawQuery("PRAGMA table_info(" + CommodityEntry.TABLE_NAME + ")", null);
        try {
            if (cursor.moveToFirst()) {
                final int indexName = cursor.getColumnIndex("name");
                do {
                    String name = cursor.getString(indexName);
                    if (CommodityEntry.COLUMN_QUOTE_FLAG.equals(name)) {
                        hasColumnQuoteFlag = true;
                        break;
                    }
                } while (cursor.moveToNext());
            }
        } finally {
            cursor.close();
        }

        if (!hasColumnQuoteFlag) {
            // Restore the currency code column that was deleted in v19.
            String sqlCommodityFlag = "ALTER TABLE " + CommodityEntry.TABLE_NAME
                + " ADD COLUMN " + CommodityEntry.COLUMN_QUOTE_FLAG + " tinyint default 0";
            try {
                db.execSQL(sqlCommodityFlag);
            } catch (SQLException e) {
                Timber.e(e);
            }
        }

        try {
            DatabaseHolder holder = new DatabaseHolder(context, db);
            importCommodities(holder);
        } catch (SAXException | ParserConfigurationException | IOException e) {
            String msg = "Error loading currencies into the database";
            Timber.e(e, msg);
            throw new SQLiteException(msg, e);
        }
    }
}

File: app/src/main/java/org/gnucash/android/db/MigrationHelper.kt
/*
 * Copyright (c) 2014 - 2015 Ngewi Fet <ngewif@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.gnucash.android.db

import android.content.Context
import android.database.DatabaseUtils.sqlEscapeString
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import androidx.annotation.VisibleForTesting
import org.gnucash.android.R
import org.gnucash.android.db.DatabaseSchema.AccountEntry
import org.gnucash.android.db.DatabaseSchema.BudgetAmountEntry
import org.gnucash.android.db.DatabaseSchema.CommodityEntry
import org.gnucash.android.db.DatabaseSchema.ScheduledActionEntry
import org.gnucash.android.db.DatabaseSchema.SplitEntry
import org.gnucash.android.db.DatabaseSchema.TransactionEntry
import org.gnucash.android.importer.CommoditiesXmlHandler
import org.gnucash.android.model.Commodity
import org.xml.sax.InputSource
import org.xml.sax.SAXException
import timber.log.Timber
import java.io.IOException
import java.io.InputStream
import javax.xml.parsers.ParserConfigurationException
import javax.xml.parsers.SAXParserFactory

/**
 * Collection of helper methods which are used during database migrations
 *
 * @author Ngewi Fet <ngewif@gmail.com>
 */
object MigrationHelper {
    /**
     * Imports commodities into the database from XML resource file
     */
    @VisibleForTesting
    @Throws(SAXException::class, ParserConfigurationException::class, IOException::class)
    fun importCommodities(holder: DatabaseHolder) {
        val parserFactory = SAXParserFactory.newInstance()
        val parser = parserFactory.newSAXParser()
        val reader = parser.xmlReader

        val inputStream: InputStream = holder.context.resources
            .openRawResource(R.raw.iso_4217_currencies)

        /* Create handler to handle XML Tags ( extends DefaultHandler ) */
        val handler = CommoditiesXmlHandler(holder)

        reader.contentHandler = handler
        reader.parse(InputSource(inputStream))
    }

    fun migrate(context: Context, db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion < 16) {
            migrateTo16(db)
        }
        if (oldVersion < 17) {
            migrateTo17(db)
        }
        if (oldVersion < 18) {
            migrateTo18(db)
        }
        if (oldVersion < 19) {
            migrateTo19(db)
        }
        if ((oldVersion >= 19) && (oldVersion < 21)) {
            migrateTo21(db)
        }
        if (oldVersion < 23) {
            migrateTo23(context, db)
        }
        if (oldVersion < 24) {
            migrateTo24(db)
        }
        if (oldVersion < 25) {
            migrateTo25(db)
        }
        if (oldVersion < 27) {
            migrateTo27(db)
        }
    }

    /**
     * Upgrade the database to version 16.
     *
     * @param db the database.
     */
    private fun migrateTo16(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 16")

        val sqlAddQuoteSource = "ALTER TABLE " + CommodityEntry.TABLE_NAME +
                " ADD COLUMN " + CommodityEntry.COLUMN_QUOTE_SOURCE + " varchar(255)"
        val sqlAddQuoteTZ = "ALTER TABLE " + CommodityEntry.TABLE_NAME +
                " ADD COLUMN " + CommodityEntry.COLUMN_QUOTE_TZ + " varchar(100)"

        db.execSQL(sqlAddQuoteSource)
        db.execSQL(sqlAddQuoteTZ)
    }

    /**
     * Upgrade the database to version 17.
     *
     * @param db the database.
     */
    private fun migrateTo17(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 17")

        val sqlAddBudgetNotes = "ALTER TABLE " + BudgetAmountEntry.TABLE_NAME +
                " ADD COLUMN " + BudgetAmountEntry.COLUMN_NOTES + " text"

        db.execSQL(sqlAddBudgetNotes)
    }

    /**
     * Upgrade the database to version 18.
     *
     * @param db the database.
     */
    private fun migrateTo18(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 18")

        val sqlAddNotes = ("ALTER TABLE " + AccountEntry.TABLE_NAME
                + " ADD COLUMN " + AccountEntry.COLUMN_NOTES + " text")
        val sqlAddBalance = ("ALTER TABLE " + AccountEntry.TABLE_NAME
                + " ADD COLUMN " + AccountEntry.COLUMN_BALANCE + " varchar(255)")
        val sqlAddClearedBalance = ("ALTER TABLE " + AccountEntry.TABLE_NAME
                + " ADD COLUMN " + AccountEntry.COLUMN_CLEARED_BALANCE + " varchar(255)")
        val sqlAddNoClosingBalance = ("ALTER TABLE " + AccountEntry.TABLE_NAME
                + " ADD COLUMN " + AccountEntry.COLUMN_NOCLOSING_BALANCE + " varchar(255)")
        val sqlAddReconciledBalance = ("ALTER TABLE " + AccountEntry.TABLE_NAME
                + " ADD COLUMN " + AccountEntry.COLUMN_RECONCILED_BALANCE + " varchar(255)")

        db.execSQL(sqlAddNotes)
        db.execSQL(sqlAddBalance)
        db.execSQL(sqlAddClearedBalance)
        db.execSQL(sqlAddNoClosingBalance)
        db.execSQL(sqlAddReconciledBalance)
        DatabaseHelper.createResetBalancesTriggers(db)
    }

    /**
     * Upgrade the database to version 19.
     *
     * @param db the database.
     */
    private fun migrateTo19(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 19")

        // Fetch list of accounts with mismatched currencies.
        val sqlAccountCurrencyWrong =
            "SELECT DISTINCT a." + AccountEntry.COLUMN_CURRENCY + ", a." + AccountEntry.COLUMN_COMMODITY_UID + ", c." + CommodityEntry.COLUMN_UID +
                    " FROM " + AccountEntry.TABLE_NAME + " a, " + CommodityEntry.TABLE_NAME + " c" +
                    " WHERE a." + AccountEntry.COLUMN_CURRENCY + " = c." + CommodityEntry.COLUMN_MNEMONIC +
                    " AND (c." + CommodityEntry.COLUMN_NAMESPACE + " = " + sqlEscapeString(Commodity.COMMODITY_CURRENCY) +
                    " OR c." + CommodityEntry.COLUMN_NAMESPACE + " = " + sqlEscapeString(Commodity.COMMODITY_ISO4217) + ")" +
                    " AND a." + AccountEntry.COLUMN_COMMODITY_UID + " != c." + CommodityEntry.COLUMN_UID
        val accountsWrong = mutableListOf<AccountCurrency>()
        db.rawQuery(sqlAccountCurrencyWrong, null).forEach { cursor ->
            val currencyCode = cursor.getString(0)
            val commodityUIDOld = cursor.getString(1)
            val commodityUIDNew = cursor.getString(2)
            accountsWrong.add(AccountCurrency(currencyCode, commodityUIDOld, commodityUIDNew))
        }

        // Update with correct commodities.
        for (accountWrong in accountsWrong) {
            val sql = "UPDATE " + AccountEntry.TABLE_NAME +
                    " SET " + AccountEntry.COLUMN_COMMODITY_UID + " = ${sqlEscapeString(accountWrong.commodityUIDNew)}" +
                    " WHERE " + AccountEntry.COLUMN_CURRENCY + " = ${sqlEscapeString(accountWrong.currencyCode)}" +
                    " AND " + AccountEntry.COLUMN_COMMODITY_UID + " = ${sqlEscapeString(accountWrong.commodityUIDOld)}"
            db.execSQL(sql)
        }
    }

    /**
     * Upgrade the database to version 21.
     *
     * @param db the database.
     */
    private fun migrateTo21(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 21")

        if (db.hasTableColumn(AccountEntry.TABLE_NAME, AccountEntry.COLUMN_CURRENCY)) {
            return
        }

        // Restore the currency code column that was deleted in v19.
        val sqlAccountCurrency = "ALTER TABLE " + AccountEntry.TABLE_NAME +
                " ADD COLUMN " + AccountEntry.COLUMN_CURRENCY + " varchar(255)"
        try {
            db.execSQL(sqlAccountCurrency)
        } catch (e: SQLException) {
            Timber.e(e)
        }

        // Restore the currency code column.
        val sql = "ALTER TABLE " + TransactionEntry.TABLE_NAME +
                " ADD COLUMN " + TransactionEntry.COLUMN_CURRENCY + " varchar(255)"
        try {
            db.execSQL(sql)
        } catch (e: SQLException) {
            Timber.e(e)
        }
    }

    /**
     * Upgrade the database to version 23.
     *
     * @param context the context.
     * @param db      the database.
     */
    private fun migrateTo23(context: Context, db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 23")

        if (!db.hasTableColumn(CommodityEntry.TABLE_NAME, CommodityEntry.COLUMN_QUOTE_FLAG)) {
            // Restore the currency code column that was deleted in v19.
            val sql = "ALTER TABLE " + CommodityEntry.TABLE_NAME +
                    " ADD COLUMN " + CommodityEntry.COLUMN_QUOTE_FLAG + " tinyint default 0"
            try {
                db.execSQL(sql)
            } catch (e: SQLException) {
                Timber.e(e)
            }
        }

        try {
            val holder = DatabaseHolder(context, db)
            importCommodities(holder)
        } catch (e: SAXException) {
            val msg = "Error loading currencies into the database"
            Timber.e(e, msg)
            throw SQLiteException(msg, e)
        } catch (e: ParserConfigurationException) {
            val msg = "Error loading currencies into the database"
            Timber.e(e, msg)
            throw SQLiteException(msg, e)
        } catch (e: IOException) {
            val msg = "Error loading currencies into the database"
            Timber.e(e, msg)
            throw SQLiteException(msg, e)
        }
    }

    /**
     * Upgrade the database to version 24.
     *
     * @param db the database.
     */
    private fun migrateTo24(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 24")

        if (!db.hasTableColumn(AccountEntry.TABLE_NAME, AccountEntry.COLUMN_TEMPLATE)) {
            val sql = "ALTER TABLE " + AccountEntry.TABLE_NAME +
                    " ADD COLUMN " + AccountEntry.COLUMN_TEMPLATE + " tinyint default 0"
            db.execSQL(sql)
        }
        if (!db.hasTableColumn(
                SplitEntry.TABLE_NAME,
                SplitEntry.COLUMN_SCHEDX_ACTION_ACCOUNT_UID
            )
        ) {
            val sql = "ALTER TABLE " + SplitEntry.TABLE_NAME +
                    " ADD COLUMN " + SplitEntry.COLUMN_SCHEDX_ACTION_ACCOUNT_UID + " varchar(255)"
            db.execSQL(sql)
        }
    }

    /**
     * Upgrade the database to version 25.
     *
     * @param db the database.
     */
    private fun migrateTo25(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 25")

        if (!db.hasTableColumn(ScheduledActionEntry.TABLE_NAME, ScheduledActionEntry.COLUMN_NAME)) {
            val sql = "ALTER TABLE " + ScheduledActionEntry.TABLE_NAME +
                    " ADD COLUMN " + ScheduledActionEntry.COLUMN_NAME + " varchar(255)"
            db.execSQL(sql)
        }
    }

    /**
     * Upgrade the database to version 27.
     *
     * @param db the database.
     */
    private fun migrateTo27(db: SQLiteDatabase) {
        Timber.i("Upgrading database to version 27")

        if (!db.hasTableColumn(TransactionEntry.TABLE_NAME, TransactionEntry.COLUMN_NUMBER)) {
            val sql = "ALTER TABLE ${TransactionEntry.TABLE_NAME} ADD COLUMN ${TransactionEntry.COLUMN_NUMBER} varchar(255)"
            db.execSQL(sql)
        }
    }

    private class AccountCurrency(
        val currencyCode: String,
        val commodityUIDOld: String,
        val commodityUIDNew: String
    )
}